src/Sylius/Bundle/CoreBundle/Tests/Form/Type/CatalogPromotion/CatalogPromotionActionTypeTest.php

```diff
@@ -0,0 +1,246 @@
+        );
+
+        $this->assertSubmittingForm($this->percentageDiscount, $newFormData);
+    }
+
+    /** @test */
+    public function it_changes_action_to_percentage_discount(): void
+    {
+        $newFormData = [
+            'type' => 'percentage_discount',
+            'configuration' => [
+                'amount' => 10
+            ]
+        ];
+
+        $this->channelRepository->findAll(Argument::any())->willReturn(
+            [$this->channel->reveal()]
+        );
+
+        $this->assertSubmittingForm($this->fixedDiscount, $newFormData);
+    }
+
+    /** @test */
+    public function it_changes_action_to_percentage_discount_with_empty_amount(): void
+    {
+        $newFormData = [
+            'type' => 'percentage_discount',
+            'configuration' => [
+                'amount' => ''
+            ]
+        ];
+
+        $this->channelRepository->findAll(Argument::any())->willReturn(
+            [$this->channel->reveal()]
+        );
+
+        $this->assertSubmittingForm($this->fixedDiscount, $newFormData);
+    }
+
+    /** @test */
+    public function it_changes_action_to_fixed_discount_with_empty_amount(): void
+    {
+        $newFormData = [
+            'type' => 'fixed_discount',
+            'configuration' => [
+                'WEB_US' => [
+                    'amount' => 10
+                ]
+            ]
+        ];
+
+        $this->channelRepository->findAll(Argument::any())->willReturn(
+            [$this->channel->reveal()]
+        );
+
+        $this->assertSubmittingForm($this->percentageDiscount, $newFormData);
+    }
+
+    /** @test */
+    public function it_updates_fixed_discount_with_not_valid_amount(): void
+    {
+        $newFormData = [
+            'type' => 'fixed_discount',
+            'configuration' => [
+                'WEB_US' => [
+                    'amount' => 'Not valid amount'
+                ]
+            ]
+        ];
+
+        $this->channelRepository->findAll(Argument::any())->willReturn(
+            [$this->channel->reveal()]
+        );
+
+        $this->assertSubmittingForm($this->fixedDiscount, $newFormData);
+    }
+
+    /** @test */
+    public function it_updates_percentage_discount_with_not_valid_amount(): void
+    {
+        $newFormData = [
+            'type' => 'percentage_discount',
+            'configuration' => [
+                'amount' => 'Not valid amount'
+            ]
+        ];
+
+        $this->channelRepository->findAll(Argument::any())->willReturn(
+            [$this->channel->reveal()]
+        );
+
+        $this->assertSubmittingForm($this->percentageDiscount, $newFormData);
+    }
+
+    private function assertSubmittingForm(CatalogPromotionActionInterface $actualFormData, array $newFormData): void
+    {
+        $form = $this->factory->create(CatalogPromotionActionType::class, $actualFormData);
+        $form->submit($newFormData);
+
+        $this->assertTrue($form->isSynchronized());
```

Comment:
This seems to me to be a quite weak check of the correctness of the form. Can we check if the values are properly assigned in the array? Something like this: 
<img width="981" alt="image" src="https://user-images.githubusercontent.com/6213903/154706370-803d6f3a-df1d-4c95-bc33-e2d010b98175.png">

Reply:
Ok, I've added new assertions 😃